// @microsoft/sp-application-base/lib/index.js
/**
 * SharePoint Framework support for building and extending client-side applications.
 *
 * @packagedocumentation
 */
// Application Customizer
export { default as ApplicationCustomizerContext } from './extensibility/ApplicationCustomizerContext';
export { default as BaseApplicationCustomizer } from './extensibility/BaseApplicationCustomizer';
export { default as ApplicationAccessor } from './extensibility/ApplicationAccessor';
export { default as _DialogManagerConfiguration } from './extensibility/DialogManagerConfiguration';
export { default as _SearchQueryManager } from './extensibility/SearchQueryManager';
// Placeholders
export { default as PlaceholderContent } from './extensibility/placeholder/PlaceholderContent';
export { default as PlaceholderName } from './extensibility/placeholder/PlaceholderName';
export { default as _Placeholder } from './extensibility/placeholder/Placeholder';
export { default as PlaceholderProvider } from './extensibility/placeholder/PlaceholderProvider';
export { default as _PlaceholderManager } from './extensibility/placeholder/PlaceholderManager';
// Base application
export { default as BaseApplication } from './BaseApplication';
export { default as _OnBeforeNavigationEventName } from './navigator/OnBeforeNavigationEventName';
export { default as BaseApplicationContext } from './BaseApplicationContext';
export { default as _ApplicationManager } from './ApplicationManager';
export { default as _Navigator } from './navigator/Navigator';
export { NavigationOperation as _NavigationOperation, NavigationUnsupportedReason as _NavigationUnsupportedReason } from './navigator/INavigationResult';
export { default as _NavigationDataError, NavigationErrorCode as _NavigationErrorCode } from './navigator/NavigationDataError';
export { default as _PrefetchedDataEventArgs } from './navigator/PrefetchedDataEventArgs';
export { ApplicationLoadType as _ApplicationLoadType } from './ApplicationLoadType';
// Page Chrome
export { default as _SPThemeProvider } from './pageChrome/SPThemeProvider';
