// @microsoft/sp-component-base/lib/index.js
/**
 * SharePoint Framework support for building client-side components
 *
 * @remarks
 * This package supports development of client-side components (e.g. web parts,
 * client-side extensions, and applications) built using the SharePoint Framework.
 * It defines the base classes and interfaces for functionality that is
 * common across all client-side component types.
 *
 * @packagedocumentation
 */
export { default as BaseComponent } from './BaseComponent';
export { default as BaseComponentContext } from './BaseComponentContext';
export { default as DynamicDataProvider } from './DynamicDataProvider';
export { default as DynamicDataSourceManager } from './DynamicDataSourceManager';
export { default as DynamicProperty } from './DynamicProperty';
export { default as _LegacyThirdPartyFabricCoreLoader } from './LegacyThirdPartyFabricCoreLoader';
export { ThemeChangedEventArgs, ThemeProvider } from './themeProvider/ThemeProvider';
export { default as _ComponentBaseKillSwitches } from './common/KillSwitches';
// ShimmerFactory
export { default as _ShimmerFactory } from './ShimmerFactory';
