// @microsoft/sp-loader/lib/utilities/telemetryConstants.js
// ISPComponentLoader.loadComponent
export var loadScriptQosScenarioName = 'SPComponentLoader.loadScript';
export var loadComponentQosScenarioName = 'SPComponentLoader.loadComponent';
export var loadComponentImplQosScenarioName = 'SPComponentLoader._loadComponentImpl';
export var loadPathDependencyQosScenarioName = 'SPComponentLoader.loadPathDependency';
export var emptyComponentErrorTagName = 'EmptyComponent';
export var configureSystemJsErrorTagName = 'ConfigureSystemJs';
export var loadComponentDependenciesErrorTagName = 'LoadComponentDependencies';
export var loadPathDependenciesErrorTagName = 'LoadPathDependencies';
// Log sources
export var spStarterLogSource = { id: 'SPStarter.start' };
export var startApplicationLogSource = { id: 'SPComponentLoader.startApplication' };
export var loadComponentLogSource = { id: 'SPComponentLoader.loadComponent' };
export var loadScriptLogSource = { id: 'SPComponentLoader.loadScript' };
export var resolveAddressLogSource = { id: 'resolveAddress' };
export var componentStoreLogSource = { id: 'ComponentStore' };
export var manifestStoreLogSource = { id: 'ManifestStore' };
