// @microsoft/sp-loader/lib/systemjs/SystemJsLoader.js
// Copyright (c) Microsoft. All rights reserved.
/// <reference path="./es-module-loader.d.ts" />
import { _TraceLogger } from '@microsoft/sp-diagnostics';
import { ServiceKey, Validate, _SPFlight } from '@microsoft/sp-core-library';
import ManifestStore from '../stores/ManifestStore';
import AddressStore from '../stores/AddressStore';
import resolveAddress from '../utilities/resolveAddress';
import isCorsEnabled from '../utilities/isCorsEnabled';
import normalizeName, { normalizeFailoverPathName } from './normalizeName';
import * as telemetryConstants from '../utilities/telemetryConstants';
import ErrorBuilder from '../error/ErrorBuilder';
import strings from './SystemsJs.resx';
/**
 * Loader for SystemJS.
 * Provides the instance of SystemJS and allows methods for configuration.
 */
var SystemJsLoader = /** @class */ (function () {
    function SystemJsLoader(serviceScope) {
        this._configuredFailoverPaths = [];
        this._checkEntryPointDependenciesError = this._checkEntryPointDependenciesError.bind(this);
        this._initialize();
    }
    Object.defineProperty(SystemJsLoader.prototype, "loadEntryPointErrorProcessors", {
        get: function () {
            return [this._checkEntryPointDependenciesError];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SystemJsLoader.prototype, "loadComponentDependencyErrorProcessors", {
        get: function () {
            return [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SystemJsLoader.prototype, "loadPathDependencyErrorProcessors", {
        get: function () {
            return [];
        },
        enumerable: true,
        configurable: true
    });
    SystemJsLoader.prototype.ensure = function (manifest, module) {
        var name = normalizeName(manifest);
        return this._ensure(name, module);
    };
    /**
     * If a module hasn't been loaded with the specified name,
     * it created a new module and sets it in SystemJS
     */
    SystemJsLoader.prototype._ensure = function (name, module) {
        var system = this._system;
        if (!system.has(name)) {
            system.set(name, system.newModule(module));
        }
    };
    SystemJsLoader.prototype.delete = function (manifest) {
        return this._delete(normalizeName(manifest));
    };
    /**
     * If a module has been loaded with the specified name,
     * it delete it from SystemJS
     */
    SystemJsLoader.prototype._delete = function (name) {
        var system = this._system;
        if (system.has(name)) {
            system.delete(name);
        }
    };
    /**
     * Calls actual System.config()
     */
    SystemJsLoader.prototype.systemConfig = function (config) {
        this._originalSystemConfig.call(this._system, config);
    };
    SystemJsLoader.prototype.load = function (manifest, name, globalName) {
        var _this = this;
        return this.systemImport(normalizeName(manifest, name))
            .then(function (module) {
            var retValue = module;
            // Save "exportedModuleName" in SystemJS
            if (!name && manifest.loaderConfig.exportName) {
                retValue = module[manifest.loaderConfig.exportName];
                _this._ensure(normalizeName(manifest, manifest.loaderConfig.exportName), retValue);
            }
            return retValue;
        });
    };
    SystemJsLoader.prototype.loadFromFailoverPath = function (name) {
        return this.systemImport(normalizeFailoverPathName(name));
    };
    /**
     * Calls System.import()
     * Catches exceptions and returns a rejected promise with the error from SystemJS
     */
    SystemJsLoader.prototype.systemImport = function (name) {
        try {
            return this._system.import(name);
        }
        catch (error) {
            _TraceLogger.logError(telemetryConstants.loadComponentLogSource, error, SystemJsLoader.systemImportEventName);
            return Promise.reject(error);
        }
    };
    /** Calls System.delete() with the name matching the input manifest */
    SystemJsLoader.prototype.systemDelete = function (manifest) {
        this._system.delete(normalizeName(manifest));
    };
    /**
     * Sets the config for SystemJS. Handles global exports, renames the dependencies.
     * Also sets AddressStore with the right mapping between script and URL.
     */
    SystemJsLoader.prototype.configure = function (manifest) {
        var resources = manifest.loaderConfig.scriptResources;
        var depsMap = {};
        // Used only for failover paths. So far, this impacts react and react-dom on debug manifests.
        var globalDepsMap = {};
        for (var name_1 in resources) {
            if (resources[name_1].type === 'component') {
                var moduleConfiguration = resources[name_1];
                var resourceManifest = ManifestStore.instance.tryGetManifest(moduleConfiguration.id, moduleConfiguration.version);
                if (resourceManifest) {
                    depsMap[name_1] = normalizeName(resourceManifest);
                }
                if (moduleConfiguration.failoverPath) {
                    if (this._configuredFailoverPaths.indexOf(name_1) === -1) {
                        var normalizedName = normalizeFailoverPathName(name_1);
                        AddressStore.instance.set(normalizedName, resolveAddress(manifest, name_1));
                        globalDepsMap[name_1] = normalizedName;
                        this._configuredFailoverPaths.push(name_1);
                    }
                }
            }
            else { // path or localizedPath
                var normalizedName = void 0;
                if (name_1 === manifest.loaderConfig.entryModuleId) { // the entry point of the module
                    normalizedName = normalizeName(manifest);
                }
                else {
                    normalizedName = normalizeName(manifest, name_1);
                    depsMap[name_1] = normalizedName;
                }
                var address = resolveAddress(manifest, name_1);
                AddressStore.instance.set(normalizedName, address);
                var resource = resources[name_1];
                this._configureMetadata(normalizedName, address, resource);
            }
        }
        var packages = {};
        packages[normalizeName(manifest)] = {
            map: depsMap,
            defaultExtension: false
        };
        var config = {
            packages: packages
        };
        this.systemConfig(config);
        if (Object.keys(globalDepsMap).length) {
            this.systemConfig({
                map: globalDepsMap
            });
        }
    };
    /**
     * Returns the dependencies of a component, as detected by SystemJS when importing the module.
     * This method ignores the manifest dependencies and looks only at the JS file instead.
     * Used as a helper to find mismatching dependencies.
     *
     * @returns Array with the dependencies in a component. Empty array if dependencies could not be found.
     */
    SystemJsLoader.prototype.getDependencies = function (manifest) {
        Validate.isNotNullOrUndefined(manifest, 'manifest');
        var defined = this._system.defined; // tslint:disable-line:no-any
        // Empirically, this is saved in 'https://relative-path.invalid/<id>_<version>'
        var definedName = this.getDefinedId(manifest);
        var componentDefinition = defined[definedName];
        if (!componentDefinition) {
            return [];
        }
        return componentDefinition.deps;
    };
    /**
     * Executes the base configuration for SystemJS. It should be private but it's public
     * so STS pages can modify scriptLoad to be false.
     */
    SystemJsLoader.prototype._baseSystemConfig = function (pluginName, scriptLoad) {
        var systemConfig = {
            meta: {
                '*': {
                    loader: pluginName,
                    scriptLoad: scriptLoad
                }
            }
        };
        this.systemConfig(systemConfig);
    };
    SystemJsLoader.prototype._initialize = function () {
        this._system = this._loadSystemJs();
        this._setCustomLoader(SystemJsLoader.pluginName, this._system);
        // Create a default SystemJS config that passes everything to our loader.
        var systemConfig = {
            baseURL: SystemJsLoader._invalidBaseUrl,
            defaultJsExtensions: false
        };
        this.systemConfig(systemConfig);
        this._baseSystemConfig(SystemJsLoader.pluginName, true);
    };
    SystemJsLoader.prototype._loadSystemJs = function () {
        // Load SystemJS
        var system = DEPRECATED_UNIT_TEST
            ? require('./test/SystemJsMock')
            // tslint:disable-next-line:max-line-length
            : require('@microsoft/loader-raw-script!../../blobs/systemjs/0.19.25/dist/system.spfx');
        // Hide System.config()
        this._originalSystemConfig = system.config;
        system.config = function (config) {
            throw new Error(strings.systemConfigDisabledError);
        };
        return system;
    };
    SystemJsLoader.prototype._setCustomLoader = function (pluginName, system) {
        var loader = {
            /**
             * locate returns an address for a module name.
             * In the case of components it takes the normalized name, which is an absolute URL.
             * SystemJS gets the name as 'https://component-id.invalid/<id>', which has an entry
             * in the AddressStore.
             * If the AddressStore doesn't have an entry for the name, locate leaves the name unmodified.
             * This supports the scenario where people call System.import() with an absolute URL.
             * System.import() with a relative URL is not supported and will fail to fetch the module,
             * as the base URL is 'https://relative-path.invalid/'
             */
            locate: function (module) {
                var address = AddressStore.instance.tryGetAddress(module.name);
                if (address) {
                    return address;
                }
                else {
                    _TraceLogger.logVerbose(telemetryConstants.loadComponentLogSource, "Address not found for module name: " + module.name + ". Using it as absolute URL.");
                    // If we don't have the address mapped, assume the name IS the URL to grab it
                    return module.name;
                }
            }
        };
        // Load the loader as a system module and then set the system config. We need to do this is this order, otherwise
        // SystemJS will try to load our loader with itself, and will throw.
        system.set(pluginName, system.newModule(loader));
    };
    /**
     * Returns the id used by SystemJS in System.defined.
     * This allows to browse through SystemJS internals for additional data, like the JS dependencies.
     */
    SystemJsLoader.prototype.getDefinedId = function (manifest) {
        return SystemJsLoader._invalidBaseUrl + manifest.id + '_' + manifest.version;
    };
    /**
     * Configures the SystemJS metadata with the crossOrigin parameter,
     * and if the resource has a globalName; sets the global name configuration.
     * The crossOrigin parameter is needed for the SPFx service worker to transparently
     * read cross-domain resources.
     */
    SystemJsLoader.prototype._configureMetadata = function (normalizedName, address, resource) {
        if (_SPFlight.isEnabled(1106 /* AkamaiCrossOrigin */)) {
            var meta = {};
            // Support global names
            if (resource.globalName) {
                meta[normalizedName] = {
                    format: 'global',
                    exports: resource.globalName
                };
            }
            if (isCorsEnabled(address)) {
                (meta[normalizedName] || (meta[normalizedName] = {})).crossOrigin = 'anonymous';
            }
            var globalConfig = {
                meta: meta
            };
            this.systemConfig(globalConfig);
        }
        else {
            // Support global names
            if (resource && resource.globalName) {
                var meta = {};
                meta[normalizedName] = {
                    format: 'global',
                    exports: resource.globalName
                };
                var globalConfig = {
                    meta: meta
                };
                this.systemConfig(globalConfig);
            }
        }
    };
    /**
     * Checks if the loaded JS dependencies match with dependencies declared in the manifest.
     * This can happen when a component is re-built but the manifest hasn't been modified.
     * If missing dependencies are found, it throws with an error stating the first mismatch.
     *
      * @param manifest - Manifest to check
      * @param name - Name of the entry point. Unused in this function.
     */
    SystemJsLoader.prototype._checkEntryPointDependenciesError = function (manifest, name) {
        var dependencies = this.getDependencies(manifest);
        var resources = manifest.loaderConfig.scriptResources;
        dependencies.forEach(function (depName) {
            if (!resources[depName]) {
                throw ErrorBuilder.buildModuleHasUndeclaredDependencyError(manifest, depName);
            }
        });
        return Promise.resolve();
    };
    SystemJsLoader.serviceKey = ServiceKey.create('sp-loader:SystemJsLoader', SystemJsLoader);
    SystemJsLoader.pluginName = 'sp-loader-resolver';
    SystemJsLoader.systemImportEventName = 'System.import';
    /**
     * This is the configured base URL for SystemJS.
     * When a user tries to call SystemJS with a relative path, an error will show this as the base URL.
     *
     * Example:
     * System.import("myModule") => Error: https://relative-path.invalid/myModule not found
     */
    SystemJsLoader._invalidBaseUrl = 'https://relative-path.invalid/';
    return SystemJsLoader;
}());
export default SystemJsLoader;
