// @microsoft/sp-webpart-base/lib/core/WebPartLoadDataCollector.js
// Copyright (c) Microsoft. All rights reserved.
import { _PerformanceLogger } from '@ms/sp-telemetry';
import KillSwitches from './../common/KillSwitches';
var NOT_AVAILABLE_VALUE = 'N/A';
var WebPartLoadDataCollector = /** @class */ (function () {
    function WebPartLoadDataCollector() {
    }
    WebPartLoadDataCollector.collect = function (manifest, /* tslint:disable-line:no-any */ webPartTag, qosMonitor, pageContext, registeredInViewport) {
        var getTime = function (key) {
            return _PerformanceLogger.readComponentBreakdown(webPartTag, key);
        };
        var start = getTime('start');
        var moduleLoaded = getTime('modulesLoaded');
        var loadingDelayed = getTime('loadingDelayed');
        var inViewportLoaded = getTime('inViewportLoading');
        var initialized = getTime('init');
        var end = getTime('end');
        var isSpinnyShown = !!getTime('displaySpinner');
        var cacheHit = getTime('CacheHit');
        var cacheMiss = getTime('CacheMiss');
        var prefetechedData = getTime('prefetchedData');
        var isCacheApplicable = Boolean(cacheHit || cacheMiss);
        var webPartCacheHit = isCacheApplicable ? (cacheHit ? 'true' : 'false') : 'N/A';
        var cacheMissReason = WebPartLoadDataCollector._getCacheMissReason(cacheHit, getTime);
        var isMultiGeo = Boolean(pageContext &&
            pageContext.legacyPageContext &&
            pageContext.legacyPageContext.isMultiGeoTenant);
        var perfBreakDown = _PerformanceLogger.readFullEUPLBreakDown();
        var layoutID = NOT_AVAILABLE_VALUE;
        var dataProvider = NOT_AVAILABLE_VALUE;
        var dataProviderTime = _PerformanceLogger.now();
        var layoutKey = "layout:" /* Layout */;
        var dataProviderKey = "dataProvider:" /* DataProvider */;
        // WebParts which have different layouts and dataproviders, will have perf marks such as
        // webPartTag.dataProvider: [dataProviderName]
        // webPartTag.Layout: [Layout]
        // The logic below will extract the data provider and layout, and log them for perf measurements.
        Object.keys(perfBreakDown).forEach(function (key) {
            if (key.indexOf(webPartTag) > -1) {
                // We only have 1 layout
                if (layoutID === NOT_AVAILABLE_VALUE && key.indexOf(layoutKey) > -1) {
                    layoutID = key.split(layoutKey)[1];
                    // Focus on primary data provider
                }
                else if (key.indexOf(dataProviderKey) > -1) {
                    // Pick earliest data provider, in case there are multiple
                    if (dataProvider === NOT_AVAILABLE_VALUE || perfBreakDown[key] < dataProviderTime) {
                        dataProvider = key.split(dataProviderKey)[1];
                        dataProviderTime = perfBreakDown[key];
                    }
                }
            }
        });
        if (KillSwitches.isLogWebPartLoadRenderDoneKillSwitchActivated() || !qosMonitor.hasEnded) {
            var viewportWait = loadingDelayed ? inViewportLoaded - loadingDelayed : 0;
            var extraData = {
                alias: manifest.alias,
                dataProvider: prefetechedData ? 'Prefetch' : dataProvider,
                isMultiGeo: isMultiGeo,
                initTime: Math.floor(initialized - inViewportLoaded),
                isInternal: manifest.isInternal,
                isSpinnyShown: isSpinnyShown,
                layout: layoutID,
                manifestId: manifest.id,
                moduleLoadTime: Math.floor(moduleLoaded - start),
                mySiteCacheHit: webPartCacheHit,
                cacheMissReason: cacheMissReason,
                renderTime: Math.floor(end - initialized),
                scenarioId: _PerformanceLogger.getScenarioId(),
                viewportWait: viewportWait,
                isFullPage: _PerformanceLogger.fullPageLoad
            };
            if (KillSwitches.collectInViewportData()) {
                extraData.tenantName = pageContext && pageContext.legacyPageContext
                    ? pageContext.legacyPageContext.tenantDisplayName : 'N/A';
                extraData.insideViewport = registeredInViewport;
            }
            qosMonitor.writeSuccess(extraData);
        }
    };
    WebPartLoadDataCollector._getCacheMissReason = function (cacheHit, getTime) {
        if (cacheHit) {
            return "N/A" /* NotApplicable */;
        }
        var mySiteCacheMissOnConfgMismatch = !!getTime('CacheMissConfigMismatch');
        var mySiteCacheMissOnLateFlush = !!getTime('LateFlush');
        var cachedWebPartNotFound = !!getTime('CachedWebPartNotFound');
        var cacheExpired = !!getTime('CacheExpired');
        var cacheMissReason = "NotSpecified" /* NotSpecified */;
        if (mySiteCacheMissOnConfgMismatch) {
            cacheMissReason = "CacheConfigMissmatch" /* CacheConfigMissmatch */;
        }
        else if (mySiteCacheMissOnLateFlush) {
            cacheMissReason = "LateFlush" /* LateFlush */;
        }
        else if (cachedWebPartNotFound) {
            cacheMissReason = "CachedItemNotFound" /* CachedItemNotFound */;
        }
        else if (cacheExpired) {
            cacheMissReason = "CacheExpired" /* CacheExpired */;
        }
        return cacheMissReason;
    };
    return WebPartLoadDataCollector;
}());
export default WebPartLoadDataCollector;
