// @microsoft/sp-core-library/lib/test-api/MonitoredServiceScope.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { ServiceScope } from '../index';
/**
 * The MonitoredServiceScope should be used to monitor and audit the status of a ServiceScope.
 * This class should only be used in debugging scenarios. Simply replace the usage of
 * the ServiceScope.startNewRoot with MonitoredServiceScope.startNewMonitoredRoot to take
 * advantage of additional logging capabilities.
 *
 */
var MonitoredServiceScope = /** @class */ (function (_super) {
    __extends(MonitoredServiceScope, _super);
    function MonitoredServiceScope(parent) {
        var _this = _super.call(this, parent) || this;
        _this._children = [];
        _this._level = 0;
        _this._serviceNames = [];
        _this._scopeSequenceId = 0;
        _this._totalNumberOfScopes = 0;
        if (parent) {
            _this._totalNumberOfScopes = ++parent._totalNumberOfScopes;
            _this._scopeSequenceId = _this._totalNumberOfScopes;
            _this._level = parent._level + 1;
            parent._children.push({
                children: _this.children,
                level: _this.level,
                scopeSequenceId: _this.scopeSequenceId,
                serviceNames: _this.serviceNames
            });
        }
        return _this;
    }
    /**
     * {@inheritDoc  ServiceScope.startNewRoot}
     */
    MonitoredServiceScope.startNewMonitoredRoot = function () {
        return new MonitoredServiceScope(undefined);
    };
    /**
     * Returns the parent of the current MonitoredServiceScope, or undefined if this is a root scope.
     *
     * @returns the parent service scope
     */
    MonitoredServiceScope.prototype.getParent = function () {
        return _super.prototype.getParent.call(this);
    };
    /**
     * Constructs a new MonitoredServiceScope that is a child of the current scope.
     *
     * @remarks
     * The service scopes form a tree structure, such that when consuming a service,
     * if the key is not explicitly provided by a child scope, the parent hierarchy
     * will be consulted.
     *
     * @returns the newly created root MonitoredServiceScope
     */
    MonitoredServiceScope.prototype.startNewChild = function () {
        return new MonitoredServiceScope(this);
    };
    Object.defineProperty(MonitoredServiceScope.prototype, "children", {
        /**
         * Returns an array of the current scope's children.
         */
        get: function () {
            return this._children;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MonitoredServiceScope.prototype, "level", {
        /**
         * Returns the number of generations that the current scope is from the root scope.
         */
        get: function () {
            return this._level;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MonitoredServiceScope.prototype, "scopeSequenceId", {
        /**
         * Returns the order in which the current MonitoredScope was created
         */
        get: function () {
            return this._scopeSequenceId;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MonitoredServiceScope.prototype, "serviceNames", {
        /**
         * Returns an array of the services registered under the current scope
         */
        get: function () {
            return this._serviceNames;
        },
        enumerable: true,
        configurable: true
    });
    MonitoredServiceScope.prototype._registerService = function (serviceKey, service) {
        _super.prototype._registerService.call(this, serviceKey, service);
        this._serviceNames.push(serviceKey.name);
    };
    return MonitoredServiceScope;
}(ServiceScope));
export default MonitoredServiceScope;
