// @microsoft/rush-lib/lib/logic/versionMismatch/VersionMismatchFinderProject.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const VersionMismatchFinderEntity_1 = require("./VersionMismatchFinderEntity");
class VersionMismatchFinderProject extends VersionMismatchFinderEntity_1.VersionMismatchFinderEntity {
    constructor(project) {
        super({
            friendlyName: project.packageName,
            cyclicDependencyProjects: project.cyclicDependencyProjects,
            skipRushCheck: project.skipRushCheck
        });
        this.fileManager = project.packageJsonEditor;
        this.packageName = project.packageName;
    }
    get filePath() {
        return this.fileManager.filePath;
    }
    get allDependencies() {
        return [...this.fileManager.dependencyList, ...this.fileManager.devDependencyList];
    }
    tryGetDependency(packageName) {
        return this.fileManager.tryGetDependency(packageName);
    }
    tryGetDevDependency(packageName) {
        return this.fileManager.tryGetDevDependency(packageName);
    }
    addOrUpdateDependency(packageName, newVersion, dependencyType) {
        return this.fileManager.addOrUpdateDependency(packageName, newVersion, dependencyType);
    }
    saveIfModified() {
        return this.fileManager.saveIfModified();
    }
}
exports.VersionMismatchFinderProject = VersionMismatchFinderProject;
