// @microsoft/sp-http/lib/index.js
/**
 * Base communication layer for the SharePoint Framework
 *
 * @remarks
 * This package defines the base communication layer for
 * the SharePoint Framework.  For REST calls, it handles authentication,
 * logging, diagnostics, and batching.  It also simplifies requests by
 * adding default headers that follow the recommended best practices.
 *
 * @packagedocumentation
 */
export { default as HttpClient } from './httpClient/HttpClient';
export { default as HttpClientConfiguration } from './httpClient/HttpClientConfiguration';
export { default as HttpClientResponse } from './httpClient/HttpClientResponse';
export { default as DigestCache } from './spHttpClient/DigestCache';
export { default as SPHttpClient } from './spHttpClient/SPHttpClient';
export { default as SPHttpClientCommonConfiguration } from './spHttpClient/SPHttpClientCommonConfiguration';
export { default as SPHttpClientConfiguration } from './spHttpClient/SPHttpClientConfiguration';
export { default as SPHttpClientResponse } from './spHttpClient/SPHttpClientResponse';
export { default as SPHttpClientBatchConfiguration } from './spHttpClient/SPHttpClientBatchConfiguration';
export { default as SPHttpClientBatch } from './spHttpClient/SPHttpClientBatch';
export { default as ODataVersion } from './spHttpClient/ODataVersion';
export { default as OAuthTokenProvider } from './oauthTokenProvider/OAuthTokenProvider';
export { default as _AadTokenProviders } from './oauthTokenProvider/AadTokenProviders';
export { default as _AadConstants } from './oauthTokenProvider/AadConstants';
export { default as _GraphHttpClient } from './graphHttpClient/GraphHttpClient';
export { default as _GraphHttpClientResponse } from './graphHttpClient/GraphHttpClientResponse';
export { default as _GraphHttpClientConfiguration } from './graphHttpClient/GraphHttpClientConfiguration';
export { default as _GraphHttpClientContext } from './graphHttpClient/GraphHttpClientContext';
export { default as AadTokenProvider, TokenAcquisitionEventArgs } from './oauthTokenProvider/AadTokenProvider';
export { default as AadHttpClient } from './aadHttpClient/AadHttpClient';
export { default as AadHttpClientConfiguration } from './aadHttpClient/AadHttpClientConfiguration';
export { default as AadHttpClientFactory } from './aadHttpClient/AadHttpClientFactory';
export { default as AadTokenProviderFactory } from './oauthTokenProvider/AadTokenProviderFactory';
export { default as MSGraphClient } from './msGraphClient/MSGraphClient';
export { default as MSGraphClientFactory } from './msGraphClient/MSGraphClientFactory';
export { default as _AdalAuthContextManager } from './oauthTokenProvider/AdalAuthContextManager';
export { MsalTokenProvider } from './oauthTokenProvider/MsalTokenProvider';
export { cacheDataProviderServiceKey as _cacheDataProviderServiceKey } from './caching/CacheDataProviderService';
export { CacheManager as _CacheManager } from './caching/CacheManager';
