// @microsoft/sp-build-core-tasks/lib/configJson/configJsonReader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const node_core_library_1 = require("@microsoft/node-core-library");
const config_v2_1 = require("./interfaces/config-v2");
const V1_SCHEMA = node_core_library_1.JsonSchema.fromFile(path.join(__dirname, 'schemas', 'config.1.0.schema.json'));
const V2_SCHEMA = node_core_library_1.JsonSchema.fromFile(path.join(__dirname, 'schemas', 'config.2.0.schema.json'));
exports.defaultConfigPath = path.join(process.cwd(), 'config', 'config.json');
/**
 * Read a V1 config file and translate it to the latest schema.
 */
function _tryV1(filename, fileMetadata) {
    const result = {
        missingFile: false,
        version: '1.0',
        isLatestVersion: false
    };
    try {
        const v1Config = node_core_library_1.JsonFile.loadAndValidate(filename, V1_SCHEMA);
        const configData = {
            '$schema': config_v2_1.configJsonSchemaUrl,
            version: '2.0',
            bundles: {},
            localizedResources: v1Config.localizedResources,
            externals: v1Config.externals
        };
        if (v1Config.entries) {
            for (const entry of v1Config.entries) {
                const name = path.basename(entry.outputPath, path.extname(entry.outputPath));
                configData.bundles[name] = {
                    components: [{
                            entrypoint: entry.entry,
                            manifest: entry.manifest
                        }]
                };
            }
        }
        if (configData.localizedResources) {
            for (const resourceName of Object.getOwnPropertyNames(configData.localizedResources)) {
                const resourcePath = configData.localizedResources[resourceName];
                configData.localizedResources[resourceName] = path.posix.join('lib', resourcePath);
            }
        }
        result.configData = configData;
    }
    catch (error) {
        if (!fileMetadata.version) {
            result.readError = error;
        }
        else {
            // Failed to validate, not the right version
            return undefined;
        }
    }
    return result;
}
/**
 * Read a V2 config file. This is currently the latest config version
 */
function _tryV2(filename, fileMetadata) {
    const result = {
        missingFile: false,
        version: '2.0',
        isLatestVersion: true
    };
    try {
        const configData = node_core_library_1.JsonFile.loadAndValidate(filename, V2_SCHEMA);
        result.configData = configData;
    }
    catch (error) {
        if (fileMetadata.version === result.version) {
            result.readError = error;
        }
        else {
            // Failed to validate, not the right version
            return undefined;
        }
    }
    return result;
}
const _parseFunctions = [
    _tryV2,
    _tryV1
];
function readConfigFile(// tslint:disable-line:export-name
logVerbose, filename = exports.defaultConfigPath) {
    if (!fs.existsSync(filename)) {
        return {
            missingFile: true
        };
    }
    else {
        logVerbose(`Found config file: ${path.basename(filename)}`);
        let fileMetadata;
        try {
            // Read the file to detect basic metadata
            fileMetadata = node_core_library_1.JsonFile.load(filename);
        }
        catch (error) {
            return {
                missingFile: false,
                readError: error
            };
        }
        let result = undefined;
        for (let i = 0; i < _parseFunctions.length && !result; i++) {
            result = _parseFunctions[i](filename, fileMetadata);
        }
        return result || {
            missingFile: false,
            version: fileMetadata.version,
            isLatestVersion: false,
            readError: 'Unknown config.json file format.'
        };
    }
}
exports.readConfigFile = readConfigFile;
function readV1ConfigFile(// tslint:disable-line:export-name
logVerbose, filename = exports.defaultConfigPath) {
    if (!fs.existsSync(filename)) {
        return {
            missingFile: true
        };
    }
    else {
        logVerbose(`Found config file: ${path.basename(filename)}`);
        let result = {
            missingFile: false,
            version: '1.0',
            isLatestVersion: false
        };
        try {
            const v1Config = node_core_library_1.JsonFile.loadAndValidate(filename, V1_SCHEMA);
            result.configData = v1Config;
        }
        catch (error) {
            result = {
                missingFile: false,
                version: result.version,
                isLatestVersion: false,
                readError: error
            };
        }
        return result;
    }
}
exports.readV1ConfigFile = readV1ConfigFile;
