// @microsoft/set-webpack-public-path-plugin/lib/codeGenerator.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const os_1 = require("os");
const uglify = require("uglify-js");
/**
 * @public
 */
exports.registryVariableName = 'window.__setWebpackPublicPathLoaderSrcRegistry__';
const varName = 'publicPath';
function getSetPublicPathCode(options, emitWarning) {
    if (!options.webpackPublicPathVariable) {
        throw '"webpackPublicPathVariable" option must be defined.';
    }
    let lines = [];
    if (options.regexName) {
        // Double-escape backslashes to make them show up as single backslashes in regexps.
        const escapedRegex = options.regexName.replace(/\\/, '\\\\');
        lines = [
            `var scripts = document.getElementsByTagName('script');`
        ];
        const regexInitializationSnippet = `new RegExp('${escapeSingleQuotes(escapedRegex)}', 'i')`;
        const regexVarName = options.regexVariable;
        if (options.regexVariable) {
            lines.push(...[
                `var regex = (typeof ${regexVarName} !== 'undefined') ? ${regexVarName} : ${regexInitializationSnippet};`
            ]);
        }
        else {
            lines.push(...[
                `var regex = ${regexInitializationSnippet};`
            ]);
        }
        lines.push(...[
            `var ${varName};`,
            '',
            'if (scripts && scripts.length) {',
            '  for (var i = 0; i < scripts.length; i++) {',
            '    if (!scripts[i]) continue;',
            `    var path = scripts[i].getAttribute('src');`,
            '    if (path && path.match(regex)) {',
            `      ${varName} = path.substring(0, path.lastIndexOf('/') + 1);`,
            ...(options.preferLastFoundScript ? [] : ['      break;']),
            '    }',
            '  }',
            '}',
            '',
            `if (!${varName}) {`,
            `  for (var global in ${exports.registryVariableName}) {`,
            '    if (global && global.match(regex)) {',
            `      ${varName} = global.substring(0, global.lastIndexOf('/') + 1);`,
            ...(options.preferLastFoundScript ? [] : ['      break;']),
            '    }',
            '  }',
            '}'
        ]);
        if (options.getPostProcessScript) {
            lines.push(...[
                '',
                `if (${varName}) {`,
                `  ${options.getPostProcessScript(varName)};`,
                '}',
                ''
            ]);
        }
    }
    else {
        if (options.publicPath) {
            lines.push(...[
                `var ${varName} = '${appendSlashAndEscapeSingleQuotes(options.publicPath)}';`,
                ''
            ]);
        }
        else if (options.systemJs) {
            lines.push(...[
                `var ${varName} = window.System ? window.System.baseURL || '' : '';`,
                `if (${varName} !== '' && ${varName}.substr(-1) !== '/') ${varName} += '/';`,
                ''
            ]);
        }
        else {
            emitWarning(`Neither 'publicPath' nor 'systemJs' is defined, so the public path will not be modified`);
            return '';
        }
        if (options.urlPrefix && options.urlPrefix !== '') {
            lines.push(...[
                `${varName} += '${appendSlashAndEscapeSingleQuotes(options.urlPrefix)}';`,
                ''
            ]);
        }
        if (options.getPostProcessScript) {
            lines.push(...[
                `if (${varName}) {`,
                `  ${options.getPostProcessScript(varName)};`,
                '}',
                ''
            ]);
        }
    }
    lines.push(`${options.webpackPublicPathVariable} = ${varName};`);
    return joinLines(lines, options.linePrefix);
}
exports.getSetPublicPathCode = getSetPublicPathCode;
/**
 * /**
 * This function returns a block of JavaScript that maintains a global register of script tags.
 *
 * @param debug - If true, the code returned code is not minified. Defaults to false.
 *
 * @public
 */
function getGlobalRegisterCode(debug = false) {
    const lines = [
        '(function(){',
        `if (!${exports.registryVariableName}) ${exports.registryVariableName}={};`,
        `var scripts = document.getElementsByTagName('script');`,
        'if (scripts && scripts.length) {',
        '  for (var i = 0; i < scripts.length; i++) {',
        '    if (!scripts[i]) continue;',
        `    var path = scripts[i].getAttribute('src');`,
        `    if (path) ${exports.registryVariableName}[path]=true;`,
        '  }',
        '}',
        '})();'
    ];
    const joinedScript = joinLines(lines);
    if (debug) {
        return `${os_1.EOL}${joinedScript}`;
    }
    else {
        const minifyOutput = uglify.minify(joinedScript, {
            compress: {
                dead_code: true
            }
        });
        return `${os_1.EOL}${minifyOutput.code}`;
    }
}
exports.getGlobalRegisterCode = getGlobalRegisterCode;
function joinLines(lines, linePrefix) {
    return lines.map((line) => {
        if (!!line) {
            return `${linePrefix || ''}${line}`;
        }
        else {
            return line;
        }
    }).join(os_1.EOL).replace(new RegExp(`${os_1.EOL}${os_1.EOL}+`, 'g'), `${os_1.EOL}${os_1.EOL}`);
}
function escapeSingleQuotes(str) {
    if (str) {
        return str.replace('\'', '\\\'');
    }
    else {
        return undefined;
    }
}
function appendSlashAndEscapeSingleQuotes(str) {
    if (str && str.substr(-1) !== '/') {
        str = str + '/';
    }
    return escapeSingleQuotes(str);
}
