// @microsoft/sp-http/lib/graphHttpClient/GraphHttpClientContext.js
import { ServiceKey } from '@microsoft/sp-core-library';
/**
 * This class is useful for getting configuration properties to the GraphHttpClient
 * @internal
 */
var GraphHttpClientContext = /** @class */ (function () {
    function GraphHttpClientContext(serviceScope) {
        // Service Scope Pattern
    }
    /**
     * The initialize function should be called before the application/components
     * have a chance to make a GraphHttpClient request.
     */
    GraphHttpClientContext.prototype.initialize = function (webServerRelativeUrl, graphEndpointUrl, webAbsoluteUrl) {
        if (this._isInitialized) {
            throw new Error('GraphHttpClientContext has already been initialized.');
        }
        this._isInitialized = true;
        this._webServerRelativeUrl = webServerRelativeUrl;
        this._graphEndpointUrl = graphEndpointUrl;
        this._webAbsoluteUrl = webAbsoluteUrl;
    };
    /**
     * Gets the Microsoft Graph endpoint for the current environment. This value is set by
     * the initialize function. If the value is not set, this function will throw.
     */
    GraphHttpClientContext.prototype.getGraphEndpointUrl = function () {
        if (!this._isInitialized) {
            throw new Error('This operation cannot be performed until the GraphHttpClientContext is initialized.');
        }
        return this._graphEndpointUrl;
    };
    /**
     * Gets the webServerRelative url of the current page. This value is set by
     * the initialize function. If the value is not set, this function will throw.
     */
    GraphHttpClientContext.prototype.getWebServerRelativeUrl = function () {
        if (!this._isInitialized) {
            throw new Error('This operation cannot be performed until the GraphHttpClientContext is initialized.');
        }
        return this._webServerRelativeUrl;
    };
    /**
   * Gets the webAbsoluteUrl of the current page. This value is set by
   * the initialize function. If the value is not set, this function will throw.
   */
    GraphHttpClientContext.prototype.getWebAbsoluteUrl = function () {
        if (!this._isInitialized) {
            throw new Error('This operation cannot be performed until the GraphHttpClientContext is initialized.');
        }
        return this._webAbsoluteUrl;
    };
    /**
     * The service key for GraphHttpClientContext.
     */
    GraphHttpClientContext.serviceKey = ServiceKey.create('sp-http:GraphHttpClientContext', GraphHttpClientContext);
    return GraphHttpClientContext;
}());
export default GraphHttpClientContext;
