// @microsoft/sp-application-base/lib/navigator/NavigationDataError.js
// Copyright (c) Microsoft. All rights reserved.
import * as tslib_1 from "tslib";
import { sealed } from '@microsoft/decorators';
import { SPError, Text, Validate } from '@microsoft/sp-core-library';
import { Killswitches } from '../common/Killswitches';
/**
 * Web part error codes.
 *
 * @internal
 */
export var NavigationErrorCode;
(function (NavigationErrorCode) {
    /**
     * Failure to fetch navigation data.
     */
    NavigationErrorCode[NavigationErrorCode["fetchData"] = 0] = "fetchData";
    /**
     * Failure to parse navigation data.
     */
    NavigationErrorCode[NavigationErrorCode["parseData"] = 1] = "parseData";
    /**
     * Unknown failure
     */
    NavigationErrorCode[NavigationErrorCode["unknown"] = 2] = "unknown";
})(NavigationErrorCode || (NavigationErrorCode = {}));
/**
 * Represents the contract of a Navigation Data Error.
 *
 * @internal
 */
var NavigationDataError = /** @class */ (function (_super) {
    tslib_1.__extends(NavigationDataError, _super);
    function NavigationDataError(errorCode, errorMessage, innerError, isExpected, errorPayload) {
        if (isExpected === void 0) { isExpected = false; }
        var _this = _super.call(this, NavigationErrorCode[errorCode], errorMessage, errorPayload) || this;
        // Manually set the prototype, as we can no longer extend built-in classes like Error, Array, Map, etc
        // tslint:disable-next-line:max-line-length
        // [https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work](https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work)
        //
        // Note: the prototype must also be set on any classes which extend this one
        _this.__proto__ = NavigationDataError_1.prototype; // tslint:disable-line:no-any
        _this._errorId = errorCode;
        _this.innerError = innerError;
        _this._isExpected = isExpected;
        return _this;
    }
    NavigationDataError_1 = NavigationDataError;
    /**
     * Utility function to create an error response.
     * @param errorCode - Represents the navigation error code.
     * @param innerError - Represents the inner error.
     *
     * @returns An error object.
     */
    NavigationDataError.createNavigationError = function (errorCode, innerError) {
        Validate.isNotNullOrUndefined(innerError, 'innerError');
        if (Killswitches.isNavigationDataErrorKSActive()) {
            return innerError;
        }
        var isExpected = NavigationDataError_1._isExpectedError(errorCode);
        var errorTemplate = NavigationDataError_1._errorTemplate(errorCode);
        var errorMessage = Text.format(errorTemplate, innerError.message);
        return new NavigationDataError_1(errorCode, errorMessage, innerError, isExpected);
    };
    NavigationDataError._isExpectedError = function (errorCode) {
        return errorCode === NavigationErrorCode.parseData;
    };
    NavigationDataError._errorTemplate = function (errorCode) {
        return NavigationDataError_1._errorCodeToTemplateMap[errorCode] || 'Failed to navigate. Error: {0}';
    };
    Object.defineProperty(NavigationDataError.prototype, "id", {
        /**
         * Gets the error Id.
         */
        get: function () {
            return this._errorId;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavigationDataError.prototype, "category", {
        /**
         * Gets the error category.
         */
        get: function () {
            return 'NavigationDataError';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavigationDataError.prototype, "isExpected", {
        /**
         * Gets a value indicating whether the error is expected.
         */
        get: function () {
            return this._isExpected;
        },
        enumerable: true,
        configurable: true
    });
    var NavigationDataError_1;
    NavigationDataError._errorCodeToTemplateMap = new Map([
        [NavigationErrorCode.fetchData, 'Failed to retrieve navigation data. Error: {0}'],
        [NavigationErrorCode.parseData, 'Failed to parse navigation data. Error: {0}']
    ]);
    NavigationDataError = NavigationDataError_1 = tslib_1.__decorate([
        sealed
    ], NavigationDataError);
    return NavigationDataError;
}(SPError));
export default NavigationDataError;
