// @microsoft/sp-core-library/dist/test-api/MockTimeProvider.js
/**
 * This is a test implementation of ITimeProvider that allows unit tests
 * to simulate the system clock.
 */
var MockTimeProvider = /** @class */ (function () {
    function MockTimeProvider(serviceScope) {
        // Pick an arbitrary starting point of 9am on Jan 1, 2010
        this.currentDate = new Date(2010, 1, 1, 9);
        // Pick an arbitrary starting time of 100 seconds.
        this.currentTimestamp = 100 * 1000;
    }
    /** Implements the ITimeProvider interface. */
    MockTimeProvider.prototype.getDate = function () {
        return this.currentDate;
    };
    /** Implements the ITimeProvider interface. */
    MockTimeProvider.prototype.getTimestamp = function () {
        return this.currentTimestamp;
    };
    /**
     * Advances the simulated clock by the specified number of milliseconds.
     */
    MockTimeProvider.prototype.addMs = function (milliseconds) {
        this.currentDate = new Date(this.currentDate.valueOf() + milliseconds);
        this.currentTimestamp += milliseconds;
    };
    return MockTimeProvider;
}());
export default MockTimeProvider;
