// @microsoft/gulp-core-build-sass/lib/CSSModules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const cssModules = require("postcss-modules");
const crypto = require("crypto");
class CSSModules {
    /**
     * CSSModules includes the source file's path relative to the project root
     * as part of the class name hashing algorithm.
     * This should be configured with the setting:
     * {@link @microsoft/gulp-core-build#IBuildConfig.rootPath}
     * That is used in {@link ./SassTask#SassTask}
     * But will default the process' current working dir.
     */
    constructor(rootPath) {
        this._classMap = {};
        if (rootPath) {
            this._rootPath = rootPath;
        }
        else {
            this._rootPath = process.cwd();
        }
    }
    getPlugin() {
        return cssModules({
            getJSON: this.saveJson.bind(this),
            generateScopedName: this.generateScopedName.bind(this)
        });
    }
    getClassMap() {
        return this._classMap;
    }
    saveJson(cssFileName, json) {
        this._classMap = json;
    }
    generateScopedName(name, fileName, css) {
        const fileBaseName = path.relative(this._rootPath, fileName);
        const safeFileBaseName = fileBaseName.replace(/\\/g, '/');
        const hash = crypto.createHmac('sha1', safeFileBaseName)
            .update(css)
            .digest('hex')
            .substring(0, 8);
        return `${name}_${hash}`;
    }
}
exports.default = CSSModules;
