// @microsoft/sp-loader/lib/DeveloperTools/Stores/BaseStore.js
/**
 * @file BaseStore.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 */
/**
 * Contains common implementation for store
 */
var BaseStore = /** @class */ (function () {
    /**
     * Initializes a new instance of BaseStore
     */
    function BaseStore() {
        this._isDisposed = false;
        this._listeners = [];
        this._isDisposed = false;
    }
    /**
     * Add a listener to this store
     * @param listener - the listener callback
     */
    BaseStore.prototype.addListener = function (listener) {
        if (!this._isDisposed) {
            this._listeners.push(listener);
        }
    };
    /**
     * Removes a listener from the store
     * @param listener - the listener callback
     */
    BaseStore.prototype.removeListener = function (listener) {
        if (!this._isDisposed) {
            var listenerIdx = this._listeners.indexOf(listener);
            if (listenerIdx > -1) {
                this._listeners.splice(listenerIdx, 1);
            }
        }
    };
    /**
     * Removes all the registered listeners
     */
    BaseStore.prototype.removeAllListeners = function () {
        this._listeners = [];
    };
    /**
     * Removes a listener from the store
     * @param listener - the listener callback
     *
     * @return the count of listeners tied to this store
     */
    BaseStore.prototype.count = function () {
        return !this._isDisposed ? this._listeners.length : 0;
    };
    /**
     * Invokes all listeners tied to this store
     */
    BaseStore.prototype.emitChange = function () {
        if (!this._isDisposed) {
            this._listeners.forEach(function (listener) { return listener(); });
        }
    };
    /**
     * Disposes all listeners
     */
    BaseStore.prototype.dispose = function () {
        if (!this._isDisposed) {
            this._listeners.splice(0);
            this._listeners = [];
            this._isDisposed = true;
        }
    };
    return BaseStore;
}());
export default BaseStore;
