// @microsoft/rush-stack-compiler-3.3/lib/shared/index.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This package is used with
 * [\@microsoft/rush-stack](https://www.npmjs.com/package/\@microsoft/rush-stack)
 * to select a TypeScript compiler version.
 *
 * It provides a supported set of versions for the following components:
 * - the TypeScript compiler
 * - [tslint](https://github.com/palantir/tslint#readme)
 * - [API Extractor](https://api-extractor.com/)
 *
 * @packageDocumentation
 */
var ApiExtractorRunner_1 = require("./ApiExtractorRunner");
exports.ApiExtractorRunner = ApiExtractorRunner_1.ApiExtractorRunner;
var RushStackCompilerBase_1 = require("./RushStackCompilerBase");
exports.RushStackCompilerBase = RushStackCompilerBase_1.RushStackCompilerBase;
var StandardBuildFolders_1 = require("./StandardBuildFolders");
exports.StandardBuildFolders = StandardBuildFolders_1.StandardBuildFolders;
var TypescriptCompiler_1 = require("./TypescriptCompiler");
exports.TypescriptCompiler = TypescriptCompiler_1.TypescriptCompiler;
var TslintRunner_1 = require("./TslintRunner");
exports.TslintRunner = TslintRunner_1.TslintRunner;
var ToolPaths_1 = require("./ToolPaths");
exports.ToolPaths = ToolPaths_1.ToolPaths;
var ToolPackages_1 = require("./ToolPackages");
exports.ToolPackages = ToolPackages_1.ToolPackages;
