// @microsoft/rush-stack-compiler-3.3/lib/shared/StandardBuildFolders.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const SRC_FOLDER_NAME = 'src';
const LIB_FOLDER_NAME = 'lib';
const DIST_FOLDER_NAME = 'dist';
const TEMP_FOLDER_NAME = 'temp';
/**
 * @beta
 */
class StandardBuildFolders {
    constructor(projectFolderPath) {
        this._projectFolderPath = projectFolderPath;
    }
    get projectFolderPath() {
        return this._projectFolderPath;
    }
    get srcFolderPath() {
        if (!this._srcFolderPath) {
            this._srcFolderPath = path.join(this._projectFolderPath, SRC_FOLDER_NAME);
        }
        return this._srcFolderPath;
    }
    get libFolderPath() {
        if (!this._libFolderPath) {
            this._libFolderPath = path.join(this._projectFolderPath, LIB_FOLDER_NAME);
        }
        return this._libFolderPath;
    }
    get distFolderPath() {
        if (!this._distFolderPath) {
            this._distFolderPath = path.join(this._projectFolderPath, DIST_FOLDER_NAME);
        }
        return this._distFolderPath;
    }
    get tempFolderPath() {
        if (!this._tempFolderPath) {
            this._tempFolderPath = path.join(this._projectFolderPath, TEMP_FOLDER_NAME);
        }
        return this._tempFolderPath;
    }
}
exports.StandardBuildFolders = StandardBuildFolders;
