// @microsoft/sp-search-extensibility/lib/BaseSearchQueryModifier.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { BaseExtension } from '@microsoft/sp-extension-base';
/**
 * Scenario for the search query modifier.
 *
 * @beta
 */
export var SearchQueryScenario;
(function (SearchQueryScenario) {
    /**
     * Search results scenario.
     */
    SearchQueryScenario[SearchQueryScenario["SearchResults"] = 1] = "SearchResults";
    /**
     * Search suggestions scenario.
     */
    SearchQueryScenario[SearchQueryScenario["SearchSuggestions"] = 2] = "SearchSuggestions";
})(SearchQueryScenario || (SearchQueryScenario = {}));
/**
 * This is the base class that third parties should extend when implementing
 * a client-side extension that can modify the search query.
 *
 * In the component manifest, the "extensionType" should be set to "SearchQueryModifier".
 *
 * @beta
 */
var BaseSearchQueryModifier = /** @class */ (function (_super) {
    __extends(BaseSearchQueryModifier, _super);
    function BaseSearchQueryModifier() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(BaseSearchQueryModifier.prototype, "timeout", {
        /**
         * Timeout for the search query modifier in milliseconds.
         * If the execution takes longer than the timeout, the original query will be used instead of the modified one.
         *
         * @virtual
         */
        get: function () { return 10000; },
        enumerable: true,
        configurable: true
    });
    /**
     * Returns a modified search query.
     *
     * @param query - Query to modify.
     * @param scenario - Search scenario.
     * @returns - Modified query.
     *
     * @virtual
     */
    BaseSearchQueryModifier.prototype.modifySearchQuery = function (query, scenario) {
        return Promise.resolve(query);
    };
    return BaseSearchQueryModifier;
}(BaseExtension));
export default BaseSearchQueryModifier;
