// @microsoft/sp-search-extensibility/lib/SearchQueryExtensionManager.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Guid } from '@microsoft/sp-core-library';
import { _QosMonitor } from '@microsoft/sp-diagnostics';
import { _ExtensionManager } from '@microsoft/sp-extension-base';
import BaseSearchQueryModifier, { SearchQueryScenario } from './BaseSearchQueryModifier';
import SearchQueryModifierContext from './SearchQueryModifierContext';
/**
 * Manages SearchQueryModifier extensions.
 * This is used by sp-application-base:SearchQueryManager to process a search query through the modifiers.
 *
 * @internal
 */
var SearchQueryExtensionManager = /** @class */ (function () {
    function SearchQueryExtensionManager(serviceScope) {
        this._extensionManager = new _ExtensionManager(serviceScope, BaseSearchQueryModifier);
        this._extensions = [];
    }
    /**
     * {@inheritDoc _ISearchQueryExtensionManager.initializeExtensions}
     */
    SearchQueryExtensionManager.prototype.initializeExtensions = function (customActions) {
        var _this = this;
        var qosMonitor = new _QosMonitor('SearchQueryExtensionManager.initializeExtensions');
        var modifierCustomActions = this._getCustomActions(customActions)
            .filter(function (ca) { return ca.location === 'ClientSideExtension.SearchQueryModifier'; }) // VSO #770394 - Re-use the string
            .sort(function (ca1, ca2) { return (ca1.sequence || Number.MAX_VALUE) - (ca2.sequence || Number.MAX_VALUE); });
        var extensionPromises = modifierCustomActions.map(function (customAction) {
            return _this._extensionManager.createExtension(customAction.clientSideComponentId.toString(), customAction.clientSideComponentProperties, function (params) { return new SearchQueryModifierContext(params); });
        });
        return Promise.all(extensionPromises)
            .then(function (loadedExtensions) {
            _this._extensions = loadedExtensions;
            var numberOfExtensions = _this._extensions.length;
            qosMonitor.writeSuccess({ numberOfExtensions: numberOfExtensions });
            return numberOfExtensions;
        })
            .catch(function (e) {
            qosMonitor.writeUnexpectedFailure(undefined, e, { numberOfExtensions: 0 });
            throw e;
        });
    };
    /**
     * {@inheritDoc _ISearchQueryExtensionManager.getSearchQuery}
     */
    SearchQueryExtensionManager.prototype.getSearchQuery = function (queryText) {
        var _this = this;
        var result = Promise.resolve({
            queryText: queryText,
            originalQueryText: queryText
        });
        this._extensions.forEach(function (extension) {
            result = result.then(function (q) { return _this._processQuery(extension, q, queryText); });
        });
        return result.then(function (q) { return q.queryText; });
    };
    SearchQueryExtensionManager.prototype._processQuery = function (extension, query, originalQueryText) {
        var qosMonitor = new _QosMonitor('SearchQueryExtensionManager.processQuery');
        var extraData = {
            componentId: extension.componentId,
            alias: extension.manifest.alias
        };
        var timeoutPromise = new Promise(function (resolve) {
            setTimeout(function () {
                qosMonitor.writeExpectedFailure('Timeout', undefined, extraData);
                resolve(query);
            }, extension.timeout);
        });
        return Promise.race([
            timeoutPromise,
            extension.modifySearchQuery(__assign({}, query, { originalQueryText: originalQueryText }), SearchQueryScenario.SearchResults)
        ])
            .then(function (q) {
            qosMonitor.writeSuccess(extraData);
            return q;
        })
            .catch(function (e) {
            qosMonitor.writeUnexpectedFailure(undefined, e, extraData);
            throw e;
        });
    };
    SearchQueryExtensionManager.prototype._getCustomActions = function (customActions) {
        return this._getDebugCustomActions().concat(customActions);
    };
    // Minimal code. Taken from ApplicationCustomizerLoader. Should be refactored. VSO #770394
    SearchQueryExtensionManager.prototype._getDebugCustomActions = function () {
        var result = [];
        var parameters = new URL(window.location.href).searchParams;
        var parameterValue = parameters.get('customActions');
        if (parameterValue) {
            var decodedParameterValue = decodeURIComponent(parameterValue);
            var queryValue = JSON.parse(decodedParameterValue);
            if (queryValue) {
                for (var _i = 0, _a = Object.keys(queryValue); _i < _a.length; _i++) {
                    var key = _a[_i];
                    var clientSideComponentId = Guid.tryParse(key);
                    var location_1 = queryValue[key].location;
                    var sequence = queryValue[key].sequence;
                    result.push({
                        clientSideComponentId: clientSideComponentId.toString(),
                        location: location_1,
                        sequence: sequence
                    });
                }
            }
        }
        return result;
    };
    return SearchQueryExtensionManager;
}());
export default SearchQueryExtensionManager;
