// @microsoft/sp-build-core-tasks/lib/packageSolution/packageSolution/utils/readCustomFeatures.js
"use strict";
/**
 * @file readCustomFeatures.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * Reads custom feature XML from folder specified by config. This function looks for all `.rels` files in
 * the `/_rels` folder, assumes that a Feature.xml file of the same name exists in the root folder. It
 * returns an object with a list of IFile files to include in the package, as well as the paths to all the
 * feature.xml files, which is later referenced by the `AppManifest.xml.rels` file, thereby including the
 * custom feature in the package.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const constants_1 = require("../config/constants");
const getFiles_1 = require("./getFiles");
/**
 * Reads custom feature XML from folder specified by config. This function looks for all `.rels` files in
 * the `/_rels` folder, assumes that a Feature.xml file of the same name exists in the root folder. It
 * returns an object with a list of IFile files to include in the package, as well as the paths to all the
 * feature.xml files, which is later referenced by the `AppManifest.xml.rels` file, thereby including the
 * custom feature in the package.
 * @todo we allow multiple featureXML globs to be specified
 */
function readCustomFeatures(options) {
    return getCustomFeatureXmlFilenames(options.paths.featureXmlDir)
        .then((filenames) => {
        return getCustomFiles(options.paths.featureXmlDir)
            .then((files) => {
            return {
                files: files,
                customFeatureFilepath: filenames
            };
        });
    });
}
exports.default = readCustomFeatures;
/** Reads the list of feature xml filenames from a specified rels folder */
function getCustomFeatureXmlFilenames(featureXmlDir) {
    return getFiles_1.getRelativeFilePaths(path.join(featureXmlDir, constants_1.default.RelsFolder), `*${constants_1.default.RelsXml}`, 'Found custom feature file: ')
        .then((filenames) => {
        return filenames.map((filename) => filename.replace(constants_1.default.RelsExtensionRegex, ''));
    });
}
/** Loads a directory of files into IFile objects */
function getCustomFiles(featureXmlDir) {
    return getFiles_1.getFiles(featureXmlDir, path.join('**', '*.*'), 'Found custom file: ');
}
