// @microsoft/sp-page-context/lib/test-api/MockPageContextData.js
import { SPPermission } from '../index'; // <-- part of "test-api" pattern; see sp-build-internal-web/includes/test-api-pattern/README.md
/**
 * This object is useful for mocking out the PageContext
 * @internal
 */
var MockPageContextData = {
    cultureInfo: {
        currentCultureName: 'en-US',
        currentUICultureName: 'en-US',
        isRightToLeft: false
    },
    list: {
        baseTemplate: -1,
        id: '00000000-0000-4000-b000-000000000003',
        permissions: SPPermission.fullMask.value,
        serverRelativeUrl: '/sites/test/list',
        title: 'Todo List'
    },
    listItem: {
        id: -1
    },
    page: {
        socialBarEnabled: true
    },
    site: {
        absoluteUrl: 'https://wwww.contoso.com/',
        cdnPrefix: 'spfx',
        classification: 'SPX',
        correlationId: '00000000-0000-4000-b000-000000000010',
        group: {
            id: '00000000-0000-4000-b000-000000000009',
            type: 'Public'
        },
        id: '00000000-0000-4000-b000-000000000001',
        isNoScriptEnabled: false,
        recycleBinItemCount: 5,
        serverRelativeUrl: '/',
        serverRequestPath: '/test.aspx',
        sitePagesEnabled: true,
        sitePagesFeatureVersion: 4
    },
    user: {
        displayName: 'User 1',
        email: 'user1@contoso.com',
        isAnonymousGuestUser: false,
        isExternalGuestUser: true,
        loginName: 'i:0#.w|domain\\user',
        preferUserTimeZone: true
    },
    web: {
        absoluteUrl: 'https://wwww.contoso.com/sites/test',
        id: '00000000-0000-4000-b000-000000000002',
        isAppWeb: false,
        language: -1,
        languageName: 'en-US',
        logoUrl: 'https://wwww.contoso.com/sites/test/test.jpg',
        permissions: SPPermission.fullMask.value,
        serverRelativeUrl: '/sites/test/',
        templateName: 'Blog',
        title: 'Local Workbench',
        description: 'testing'
    },
    featureInfo: {
        'test1': { Version: 1, Enabled: true },
        'test2': { Version: 2, Enabled: false }
    }
};
export default MockPageContextData;
