// @microsoft/gulp-core-build-typescript/lib/TslintCmdTask.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const node_core_library_1 = require("@microsoft/node-core-library");
const RSCTask_1 = require("./RSCTask");
/**
 * @beta
 */
class TslintCmdTask extends RSCTask_1.RSCTask {
    constructor() {
        super('tslint', {
            displayAsError: false
        });
    }
    loadSchema() {
        return node_core_library_1.JsonFile.load(path.resolve(__dirname, 'schemas', 'tslint-cmd.schema.json'));
    }
    executeTask() {
        this.initializeRushStackCompiler();
        const tslintRunner = new this._rushStackCompiler.TslintRunner({
            displayAsError: this.taskConfig.displayAsError,
            fileError: this.fileError.bind(this),
            fileWarning: this.fileWarning.bind(this)
        }, this.buildFolder, this._terminalProvider);
        return tslintRunner.invoke();
    }
}
exports.TslintCmdTask = TslintCmdTask;
